WordPress/Classes/System/WordPressAppDelegate+openURL.swift

```diff
@@ -0,0 +1,132 @@
+import WordPressAuthenticator
+
+/// Custom pattern matcher used for our URL scheme
+private func ~=(pattern: String, value: URL) -> Bool {
+    return value.absoluteString.contains(pattern)
+}
+
+private extension Array where Element == URLQueryItem {
+    func value(of key: String) -> String? {
+        return self.first(where: { $0.name == key })?.value
+    }
+    func intValue(of key: String) -> Int? {
+        guard let value = value(of: key) else {
+            return nil
+        }
+        return Int(value)
+    }
+}
+
+private extension URL {
+    func queryParams() -> [URLQueryItem]? {
+        guard let components = URLComponents(url: self, resolvingAgainstBaseURL: false),
+            let queryItems = components.queryItems,
+            queryItems.count > 0 else {
+                return nil
+        }
+        return queryItems
+    }
+}
+
+@objc extension WordPressAppDelegate {
+    internal func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
+        DDLogInfo("Application launched with URL: \(url)")
+        var returnValue = false
```

Comment:
I understand the original intention here was to reduce the number of returns, but I wonder if it actually reads more clearly in Swift to have explicit returns in each situation. We could also extract the Hockey and Google handling, removing the need for the comments here.

```
DDLogInfo("Application launched with URL: \(url)")

if handleHockeyURL(url, options: options) {
    return true
}

if handleGoogleAuthURL(url, options: options) {
    return true
}

guard url.scheme == WPComScheme else {
    return false
}

// this works with our custom ~=
switch url {
case "magic-login":
    return handleMagicLogin(url: url)
case "viewpost":
    return handleViewPost(url: url)
case "viewstats":
    return handleViewStats(url: url)
case "debugging":
    return handleDebugging(url: url)
default:
    break
}

return true
```

Those methods could look something like this:

```
private func handleHockeyURL(_ url: URL, options: [UIApplication.OpenURLOptionsKey: Any]) -> Bool {
    var hockeyOptions: [String: Any] = [:]
    for (key, value) in options {
        hockeyOptions[key.rawValue] = value
    }

    if hockey.handleOpen(url, options: hockeyOptions) {
        return true
    }

    return false
}

private func handleGoogleAuthURL(_ url: URL, options: [UIApplication.OpenURLOptionsKey: Any]) -> Bool {
    if WordPressAuthenticator.shared.handleGoogleAuthUrl(url,
                                                         sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String,
                                                         annotation: options[UIApplication.OpenURLOptionsKey.annotation]) {
        return true
    }

    return false
}
```

Let me know what you think :)

Reply:
Ya, I tried to keep the port really faithful, but I think you're right that we can step out of the old design 👍